lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java

```diff
@@ -0,0 +1,185 @@
+
+        if (rd.getDefaultConfiguration()) {
+          this.defaultRuntimeName = rd.getRuntimeName().toString();
+        }
+
+        config = serializer.fromString(rd.getSerializedConfiguration().toString());
+      } catch (IOException e) {
+        throw new RuntimeException("Unable to read runtime configuarion.", e);
+      }
+
+      Injector runtimeInjector = Tang.Factory.getTang().newInjector();
+      try {
+        initializeInjector(runtimeInjector);
+      } catch (InjectionException e) {
+        throw new RuntimeException("Unable to initialize runtimes.", e);
+      }
+
+      runtimeInjector = runtimeInjector.forkInjector(config);
+      this.runtimesInjectors.put(rd.getRuntimeName().toString(), runtimeInjector);
+    }
+  }
+
+  private void initializeInjector(final Injector runtimeInjector) throws InjectionException {
+    final EventHandler<ResourceStatusEvent> statusEventHandler =
+            this.originalInjector.getNamedInstance(RuntimeParameters.ResourceStatusHandler.class);
+    runtimeInjector.bindVolatileParameter(RuntimeParameters.ResourceStatusHandler.class, statusEventHandler);
+    final EventHandler<NodeDescriptorEvent> nodeDescriptorEventHandler =
+            this.originalInjector.getNamedInstance(RuntimeParameters.NodeDescriptorHandler.class);
+    runtimeInjector.bindVolatileParameter(RuntimeParameters.NodeDescriptorHandler.class, nodeDescriptorEventHandler);
+    final EventHandler<ResourceAllocationEvent> resourceAllocationEventHandler =
+            this.originalInjector.getNamedInstance(RuntimeParameters.ResourceAllocationHandler.class);
+    runtimeInjector.bindVolatileParameter(
+            RuntimeParameters.ResourceAllocationHandler.class,
+            resourceAllocationEventHandler);
+    final EventHandler<RuntimeStatusEvent> runtimeStatusEventHandler =
+            this.originalInjector.getNamedInstance(RuntimeParameters.RuntimeStatusHandler.class);
+    runtimeInjector.bindVolatileParameter(
+            RuntimeParameters.RuntimeStatusHandler.class,
+            runtimeStatusEventHandler);
+  }
+
+  private RuntimeDefinition parseSerializedRuntimeDefinition(final String serializedRuntimeDefinition) throws
+          IOException {
+    final RuntimeDefinition rd;
+    final JsonDecoder decoder = DecoderFactory.get().
+            jsonDecoder(RuntimeDefinition.getClassSchema(), serializedRuntimeDefinition);
+    final SpecificDatumReader<RuntimeDefinition> reader = new SpecificDatumReader<>(RuntimeDefinition.class);
+    rd = reader.read(null, decoder);
+    return rd;
+  }
+
+  private Injector getInjector(final String requestedRuntimeName) {
+    String runtimeName = requestedRuntimeName;
+    if (runtimeName.isEmpty()) {
+      runtimeName = defaultRuntimeName;
+    }
+
+    return this.runtimesInjectors.get(runtimeName);
+  }
+
+  public void onNext(final ResourceLaunchEvent value) {
+    try {
+      getInjector(value.getRuntimeName()).getInstance(ResourceLaunchHandler.class).onNext(value);
```

Comment:
Agree, it will look much nicer. Two vs. one lookup doesn't matter in this case as this is control plane, even the fact that we will be able to cache handlers instead of doing tang lookup every time probably does not matter from perf point of view, but still looks much nicer.